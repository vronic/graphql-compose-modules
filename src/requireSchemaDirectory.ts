import fs from 'fs';
import { join, resolve, dirname, basename } from 'path';

interface Options {
  extensions: string[];
  include?: RegExp | ((path: string, filename: string) => boolean);
  exclude?: RegExp | ((path: string, filename: string) => boolean);
}

type RequireAstNodeKinds = 'rootType' | 'dir' | 'file';

interface RequireAstBaseNode {
  kind: RequireAstNodeKinds;
  name: string;
  absPath: string;
}

export interface RequireAstRootTypeNode extends RequireAstBaseNode {
  kind: 'rootType';
  children: {
    [key: string]: RequireAstDirNode | RequireAstFileNode;
  };
}

export interface RequireAstDirNode extends RequireAstBaseNode {
  kind: 'dir';
  children: {
    [key: string]: RequireAstDirNode | RequireAstFileNode;
  };
}

export interface RequireAstFileNode extends RequireAstBaseNode {
  kind: 'file';
  code: {
    default?: any;
  };
}

export interface RequireAstResult {
  query?: RequireAstRootTypeNode;
  mutation?: RequireAstRootTypeNode;
  subscription?: RequireAstRootTypeNode;
}

export const defaultOptions: Options = {
  extensions: ['js', 'ts'],
};

export function requireSchemaDirectory(
  m: NodeModule,
  path?: string,
  options: Options = defaultOptions
): RequireAstResult {
  // if no path was passed in, assume the equivelant of __dirname from caller
  // otherwise, resolve path relative to the equivalent of __dirname
  const schemaPath = !path ? dirname(m.filename) : resolve(dirname(m.filename), path);

  const result = {} as RequireAstResult;

  fs.readdirSync(schemaPath).forEach((filename) => {
    const absPath = join(schemaPath, filename);

    if (fs.statSync(absPath).isDirectory()) {
      const dirName = filename;
      const re = /^(query|mutation|subscription)(\.(.*))?$/i;
      const found = dirName.match(re);
      if (found) {
        const opType = found[1].toLowerCase() as keyof RequireAstResult;
        if (!result[opType])
          result[opType] = {
            kind: 'rootType',
            name: opType,
            absPath,
            children: {},
          } as RequireAstRootTypeNode;

        const subField = found[3]; // any part after dot (eg for `query.me` will be `me`)
        if (subField) {
          result[opType].children[subField] = {
            ...requireSubDirectory(m, absPath, options),
            name: subField,
            absPath,
          };
        } else {
          result[opType].children = requireSubDirectory(m, absPath, options).children;
        }
      }
    }
  });

  return result;
}

export function requireSubDirectory(
  m: NodeModule,
  path: string,
  options: Options = defaultOptions
): RequireAstDirNode {
  const result: RequireAstDirNode = {
    kind: 'dir',
    absPath: resolve(dirname(m.filename), path),
    name: basename(path),
    children: {},
  };

  // setup default options
  Object.keys(defaultOptions).forEach((prop) => {
    if (typeof options[prop] === 'undefined') {
      options[prop] = defaultOptions[prop];
    }
  });

  // get the path of each file in specified directory, append to current tree node, recurse
  fs.readdirSync(path).forEach((filename) => {
    const absPath = join(path, filename);

    const stat = fs.statSync(absPath);
    if (stat.isDirectory()) {
      // this node is a directory; recurse
      if (result.children[filename]) {
        throw new Error(
          `You have a folder and file with same name "${filename}" by the following path ${path}. Please remove one of them.`
        );
      }
      result.children[filename] = requireSubDirectory(m, absPath, options);
    } else if (stat.isFile()) {
      // this node is a file
      if (absPath !== m.filename && checkFileInclusion(absPath, filename, options)) {
        // hash node key shouldn't include file extension
        const moduleName = filename.substring(0, filename.lastIndexOf('.'));
        if (result.children[moduleName]) {
          throw new Error(
            `You have a folder and file with same name "${moduleName}" by the following path ${path}. Please remove one of them.`
          );
        }
        result.children[moduleName] = {
          kind: 'file',
          name: moduleName,
          absPath,
          code: m.require(absPath),
        };
      }
    }
  });

  return result;
}

function checkFileInclusion(absPath: string, filename: string, options: Options): boolean {
  return (
    // verify file has valid extension
    new RegExp('\\.(' + options.extensions.join('|') + ')$', 'i').test(filename) &&
    // if options.include is a RegExp, evaluate it and make sure the path passes
    !(options.include && options.include instanceof RegExp && !options.include.test(absPath)) &&
    // if options.include is a function, evaluate it and make sure the path passes
    !(
      options.include &&
      typeof options.include === 'function' &&
      !options.include(absPath, filename)
    ) &&
    // if options.exclude is a RegExp, evaluate it and make sure the path doesn't pass
    !(options.exclude && options.exclude instanceof RegExp && options.exclude.test(absPath)) &&
    // if options.exclude is a function, evaluate it and make sure the path doesn't pass
    !(
      options.exclude &&
      typeof options.exclude === 'function' &&
      options.exclude(absPath, filename)
    )
  );
}
