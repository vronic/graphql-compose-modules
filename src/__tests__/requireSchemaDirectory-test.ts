import { requireSchemaDirectory } from '../requireSchemaDirectory';

describe('requireSchemaDirectory()', () => {
  describe('Schema ../../examples/forTests/schema', () => {
    const ast = requireSchemaDirectory(module, '../../examples/forTests/schema');

    it('should return root types', () => {
      expect(Object.keys(ast)).toEqual(expect.arrayContaining(['query', 'mutation']));
    });

    it('query has proper values', () => {
      expect(ast.query).toMatchSnapshot({
        name: 'query',
        kind: 'rootType',
        absPath: expect.any(String),
        children: {
          me: {
            absPath: expect.any(String),
            children: {
              'address.city': expect.objectContaining({ kind: 'file' }),
              'address.street': expect.objectContaining({ kind: 'file' }),
              name: expect.objectContaining({ kind: 'file' }),
              index: {
                kind: 'file',
                name: 'index',
                absPath: expect.any(String),
                code: expect.any(Object),
              },
            },
          },
          index: { kind: 'file', absPath: expect.any(String), code: expect.any(Object) },
          'some.type.index': expect.any(Object),
          user: {
            absPath: expect.any(String),
            children: {
              extendedData: expect.any(Object),
              index: expect.any(Object),
            },
          },
        },
      });
    });

    it('mutation has proper values', () => {
      expect(ast.mutation).toMatchSnapshot({
        name: 'mutation',
        kind: 'rootType',
        absPath: expect.any(String),
        children: {
          auth: {
            absPath: expect.any(String),
            children: {
              index: expect.objectContaining({ kind: 'file' }),
              login: expect.objectContaining({ kind: 'file' }),
              logout: expect.objectContaining({ kind: 'file' }),
            },
          },
          user: {
            absPath: expect.any(String),
            children: {
              create: expect.objectContaining({ kind: 'file' }),
              update: expect.objectContaining({ kind: 'file' }),
            },
          },
          'logs.nested': {
            absPath: expect.any(String),
            children: {
              list: expect.objectContaining({ kind: 'file' }),
            },
          },
        },
      });
    });
  });
});
