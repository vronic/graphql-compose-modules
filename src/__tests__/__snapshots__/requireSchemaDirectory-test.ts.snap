// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requireSchemaDirectory() Schema ../../examples/forTests/schema mutation has proper values 1`] = `
Object {
  "absPath": Any<String>,
  "children": Object {
    "auth": Object {
      "absPath": Any<String>,
      "children": Object {
        "index": ObjectContaining {
          "kind": "file",
        },
        "login": ObjectContaining {
          "kind": "file",
        },
        "logout": ObjectContaining {
          "kind": "file",
        },
      },
      "kind": "dir",
      "name": "auth",
    },
    "logs.nested": Object {
      "absPath": Any<String>,
      "children": Object {
        "list": ObjectContaining {
          "kind": "file",
        },
      },
      "kind": "dir",
      "name": "logs.nested",
    },
    "user": Object {
      "absPath": Any<String>,
      "children": Object {
        "create": ObjectContaining {
          "kind": "file",
        },
        "update": ObjectContaining {
          "kind": "file",
        },
      },
      "kind": "dir",
      "name": "user",
    },
  },
  "kind": "rootType",
  "name": "mutation",
}
`;

exports[`requireSchemaDirectory() Schema ../../examples/forTests/schema query has proper values 1`] = `
Object {
  "absPath": Any<String>,
  "children": Object {
    "index": Object {
      "absPath": Any<String>,
      "code": Any<Object>,
      "kind": "file",
      "name": "index",
    },
    "me": Object {
      "absPath": Any<String>,
      "children": Object {
        "address.city": ObjectContaining {
          "kind": "file",
        },
        "address.street": ObjectContaining {
          "kind": "file",
        },
        "index": Object {
          "absPath": Any<String>,
          "code": Any<Object>,
          "kind": "file",
          "name": "index",
        },
        "name": ObjectContaining {
          "kind": "file",
        },
      },
      "kind": "dir",
      "name": "me",
    },
    "some.type.index": Any<Object>,
    "user": Object {
      "absPath": Any<String>,
      "children": Object {
        "extendedData": Any<Object>,
        "index": Any<Object>,
        "roles": Any<Object>,
      },
      "kind": "dir",
      "name": "user",
    },
  },
  "kind": "rootType",
  "name": "query",
}
`;
